src/components/DataDimension/DndContext.js

```diff
@@ -0,0 +1,165 @@
+import {
+    DndContext,
+    DragOverlay,
+    useSensor,
+    useSensors,
+    PointerSensor,
+} from '@dnd-kit/core'
+import PropTypes from 'prop-types'
+import React, { useState } from 'react'
+import DraggingItem from './DraggingItem.js'
+
+const activateAt15pixels = {
+    activationConstraint: {
+        distance: 15,
+    },
+}
+
+export const OPTIONS_PANEL = 'Sortable'
+
+const OuterDndContext = ({ children, onDragEnd, onDragStart }) => {
+    const [draggingItem, setDraggingItem] = useState(null)
+    const sensor = useSensor(PointerSensor, activateAt15pixels)
+    const sensors = useSensors(sensor)
+
+    return (
+        <DndContext
+            collisionDetection={rectIntersectionCustom}
+            onDragStart={handleDragStart}
+            onDragEnd={handleDragEnd}
+            onDragCancel={handleDragCancel}
+            sensors={sensors}
+        >
+            {children}
+            <DragOverlay dropAnimation={null}>
+                {draggingItem ? (
+                    <span className="dragOverlay">
+                        <DraggingItem item={draggingItem} />
+                    </span>
+                ) : null}
+            </DragOverlay>
+        </DndContext>
+    )
+
+    function handleDragStart({ active }) {
+        setDraggingItem(active.data.current)
+        onDragStart && onDragStart()
+    }
+
+    function handleDragCancel() {
+        setDraggingItem(null)
+    }
+
+    function handleDragEnd({ active, over }) {
+        if (
+            !over?.id ||
+            over?.data?.current?.sortable?.containerId === OPTIONS_PANEL
+        ) {
+            // dropped over non-droppable or over options panel
+            handleDragCancel()
+            return
+        }
+
+        const activeItem = {
+            id: active.id,
+            sourceAxisId: active.data.current?.sortable?.containerId,
+            sourceIndex: active.data.current?.sortable?.index,
+            data: {
+                id: active.data.current.id,
+                type: active.data.current.type,
+                label: active.data.current.label,
+            },
+        }
+        const overItem = {
+            axisId: over.data.current?.sortable?.containerId || over.id,
+            index: over.data.current?.sortable?.index,
+        }
+        onDragEnd({ activeItem, over: overItem })
+        setDraggingItem(null)
+    }
+
+    function getIntersectionRatio(entry, target) {
```

Comment:
move this and other pure helper functions to a separate file? (or just outside of the component?)

Reply:
I've moved them outside the component. When we factor out a set of dnd tools/components for reuse
, then I'll move it